Core/src/ExponentialBackoff.php

```diff
@@ -75,10 +106,24 @@ public function execute(callable $function, array $arguments = [])
         $retryAttempt = 0;
         $exception = null;
 
+        // The $arguments are passed by reference
+        // thus are modifiable before the execution starts.
+        if ($this->onExecutionStartFunction) {
+            call_user_func_array($this->onExecutionStartFunction, [&$arguments]);
+        }
+
         while (true) {
             try {
                 return call_user_func_array($function, $arguments);
             } catch (\Exception $exception) {
+                if ($this->onRetryExceptionFunction) {
+                    // The $arguments are passed by reference so that the user has the ability to modify
+                    // some elements of the request on every retry(for example headers).
+                    call_user_func_array(
+                        $this->onRetryExceptionFunction,
+                        [$exception, $retryAttempt, &$arguments]
+                    );
```

Comment:
As mentioned below, I think we should put this after `$retryAttempt++`.

Also, why are we passing the arguments by reference here AND in the function signature? The arguments should be passed normally here, so that they _can_ be modified using pass-by-reference if desired by the person who creates the function. There's no need to have it done here.

Reply:
We are passing as & because [this](https://stackoverflow.com/questions/295016/is-it-possible-to-pass-parameters-by-reference-using-call-user-func-array#:~:text=To%20pass%20by%20reference%20using,Do%20Something...%20%7D) is happening.

Reply:
Thank you! I was not aware of that requirement!
Not required, but now that we are using PHP 7.4, we can use the [spread operator](https://www.phptutorial.net/php-tutorial/php-spread-operator/)
```php
                    $onRetryExceptionFunction = $this->onRetryExceptionFunction;
                    $onRetryExceptionFunction(...[$exception, $retryAttempt, &$arguments]);
```